src/protocol/sort/bit_permutation.rs

```diff
@@ -3,83 +3,88 @@ use std::iter::{repeat, zip};
-            });
-
-        let async_multiply =
-            zip(repeat(ctx), mult_input)
-                .enumerate()
-                .map(|(i, (ctx, (x, sum)))| async move {
-                    ctx.multiply(RecordId::from(i), &x, &sum).await
-                });
-        let mut mult_output = try_join_all(async_multiply).await?;
-
-        assert_eq!(mult_output.len(), self.input.len() * 2);
-        // Generate permutation location
-        let len = mult_output.len() / 2;
-        for i in 0..len {
-            // we are subtracting "1" from the result since this protocol returns 1-index permutation whereas all other
-            // protocols expect 0-indexed permutation
-            let less_one = &mult_output[i + len] - &share_of_one;
-            mult_output[i] = less_one + &mult_output[i];
-        }
-        mult_output.truncate(len);
+impl<F: Field> ShareOfOne<F> for ProtocolContext<'_, MaliciousReplicated<F>, F> {
+    type Share = MaliciousReplicated<F>;
 
-        Ok(mult_output)
+    fn share_of_one(&self) -> Self::Share {
+        MaliciousReplicated::one(self.role(), self.prss().generate_replicated(RECORD_0))
+    }
+}
+
+/// This is an implementation of `GenBitPerm` (Algorithm 3) described in:
+/// "An Efficient Secure Three-Party Sorting Protocol with an Honest Majority"
+/// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
+/// <https://eprint.iacr.org/2019/695.pdf>.
+#[embed_doc_image("bit_permutation", "images/sort/bit_permutations.png")]
+/// Protocol to compute a secret sharing of a permutation, after sorting on just one bit.
+///
+/// At a high level, the protocol works as follows:
+/// 1. Start with a list of `n` secret shares `[x_1]` ... `[x_n]` where each is a secret sharing of either zero or one.
+/// 2. Create a vector of length `2*n` where the first `n` rows have the values `[1 - x_1]` ... `[1 - x_n]`
+/// and the next `n` rows have the value `[x_1]` ... `[x_n]`
+/// 3. Compute a new vector of length `2*n` by computing the running sum of the vector from step 2.
+/// 4. Compute another vector of length `2*n` by multipling the vectors from steps 2 and 3 element-wise.
+/// 5. Compute the final output, a vector of length `n`. Each element `i` in this output vector is the sum of
+/// the elements at index `i` and `i+n` from the vector computed in step 4.
+///
+/// ![Bit Permutation steps][bit_permutation]
+/// ## Panics
+/// In case the function is unable to get double size of output from multiplication step, the code will panic
+///
+/// ## Errors
+/// It will propagate errors from multiplication protocol.
+pub async fn bit_permutation<'a, F: Field, S: SecretSharing<F> + Clone>(
+    ctx: ProtocolContext<'a, S, F>,
+    input: &[S],
+) -> Result<Vec<S>, BoxError> 
+where ProtocolContext<'a, S, F>: SecureMul<F, Share = S> + ShareOfOne<F, Share = S>{
+    let share_of_one = ctx.share_of_one();
+
+    let mult_input = input
+        .iter()
+        .map(|x: &S| &share_of_one - x)
```

Comment:
cannot subtract `&S` from `&S`. consider extending the `where` clause, but there might be an alternative better way to express this requirement: `, &S: Sub<&S>`

Reply:
For now, we are cloning share_of_one to take care of this.